### ISSUE TITLE
Setcap in docker build causes un-runnable container with docker defaults

### ISSUE THREAD

**Issue by schallert** [m3db/m3#1671]
In #1666 we added support for M3DB raising its own process limits. As part of this, we `setcap`'d the `m3dbnode` binary inside of the container. Doing so at build-time causes the image to fail to start unless the user specifically adds the capability:

v0.9.4 vanilla `docker run`:
```
$ docker run --rm quay.io/m3db/m3dbnode:v0.9.4
2019/05/29 01:29:07 Go Runtime version: go1.10.8
```

v0.9.5 vanilla `docker run`:
```
$ docker run --rm quay.io/m3db/m3dbnode:v0.9.5
standard_init_linux.go:207: exec user process caused "operation not permitted"
```

v0.9.5 `docker run` allowing caps:
```
$ docker run --cap-add SYS_RESOURCE --rm quay.io/m3db/m3dbnode:v0.9.5
2019/05/29 01:31:06 Go Runtime version: go1.10.8
```

Making our docker image crash on startup unless the user debugs the capability seems pretty aggressive. I believe we should either `setcap` at runtime or see if we even need to `setcap` the binary at all and can just do things at process runtime (I'm looking into this).

Regardless, I think we should revert the setcap change (at least within the build, we can leave the part where the dbnode tries to raise its own limit) until we test this out further.

**Comment by schallert** [comment]
cc @robskillington

**Comment by schallert** [comment]
As some inspiration in the problem space, here's how vault (which wants `CAP_IPC_LOCK`) [does it](https://github.com/hashicorp/docker-vault/blob/master/0.X/docker-entrypoint.sh):

1. At runtime (Docker entrypoint), `setcap` the binary
2. Attempt to run `vault version` to see if the binary can successfully be exec'd
3. If it starts, run the binary as is. If not, remove the caps and then run

**Comment by schallert** [comment]
More context on this in this issue comment: https://github.com/m3db/m3/pull/1672#issuecomment-496777131

TL;DR can't exec a binary with a cap set if you don't have that cap in your bounding set. Docker [lists](https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities) their default set, `SYS_RESOURCE` isn't in it.
```
$ ./m3dbnode 2>&1 | head -n1
2019/05/29 04:11:50 Go Runtime version: go1.10.8

$ capsh --drop=cap_sys_resource --user=mschalle -- -c '/home/mschalle/m3dbnode'
/bin/bash: /home/mschalle/m3dbnode: Operation not permitted

# un-setcap
sudo setcap cap_sys_resource-ep m3dbnode

# all good
$ capsh --drop=cap_sys_resource --user=mschalle -- -c '/home/mschalle/m3dbnode' 2>&1 | head -n1
2019/05/29 04:19:36 Go Runtime version: go1.10.8
```

Reading through the [kernel docs](http://man7.org/linux/man-pages/man7/capabilities.7.html), I think  `Safety checking for capability-dumb binaries` is the relevant section:
```
       A capability-dumb binary is an application that has been marked to
       have file capabilities, but has not been converted to use the
       libcap(3) API to manipulate its capabilities.  (In other words, this
       is a traditional set-user-ID-root program that has been switched to
       use file capabilities, but whose code has not been modified to under‐
       stand capabilities.)  For such applications, the effective capability
       bit is set on the file, so that the file permitted capabilities are
       automatically enabled in the process effective set when executing the
       file.  The kernel recognizes a file which has the effective capabil‐
       ity bit set as capability-dumb for the purpose of the check described
       here.

       When executing a capability-dumb binary, the kernel checks if the
       process obtained all permitted capabilities that were specified in
       the file permitted set, after the capability transformations
       described above have been performed.  (The typical reason why this
       might not occur is that the capability bounding set masked out some
       of the capabilities in the file permitted set.)  If the process did
       not obtain the full set of file permitted capabilities, then
       execve(2) fails with the error EPERM.  This prevents possible secu‐
       rity risks that could arise when a capability-dumb application is
       executed with less privilege that it needs.  Note that, by defini‐
       tion, the application could not itself recognize this problem, since
       it does not employ the libcap(3) API.
```

I think since we have the effective flag set but not `SYS_RESOURCE` in the bounding set of the exec'ing process (Docker, or shell in examples above) these rules apply and `execve` fails.

Going forward, I think we can either take a similar approach to vault or when M3DB starts up attempt to use the `libcap` APIs to gain the `SYS_RESOURCE` capability and _then_ try to `setrlimit`. Not sure which we should go with yet, both should be tested once we get past the current issue.

**Comment by schallert** [comment]
Keeping this open until we figure out a plan going forward.

**Comment by schallert** [comment]
Recapping some of the things I've tried and what it sounds like will be the approach we go with.

The "just syscall our way out of it" approach isn't going to work if the process doesn't start up with `SYS_RESOURCE` permitted. From the Kernel docs on capabilities:

> The new permitted set must be a subset of the existing permitted
set (i.e., it is not possible to acquire permitted capabilities
that the thread does not currently have).

Furthermore:
```
           P'(permitted)   = (P(inheritable) & F(inheritable)) |
                             (F(permitted) & P(bounding)) | P'(ambient)
```

Since Kubernetes & Docker don't yet support ambient caps (https://github.com/kubernetes/kubernetes/issues/56374 & https://github.com/moby/moby/issues/8460), the only way to support running with SYS_RESOURCE for a non-root user seems to be if the image came with the dbnode binary `setcap`'d. _However_, that would cause the image to be unable to start for users that don't know to add `SYS_RESOURCE` to the cap set.

Given all of that, here's the plan:

1. The default docker image will remain as-is, and we'll modify the dbnode warning messages to link to docs recommending `SYS_RESOURCE`. If a user is running their container as root, Docker won't give them SYS_RESOURCE by default anyway, they'd have to use `--cap-add SYS_RESOURCE`: [docs](https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities). Combining that with this from the kernel docs, if you're root in a container you won't have `SYS_RESOURCE` even if the file has it if Docker didn't give you it:

> If the real or effective user ID of the process is 0 (root), then
the file inheritable and permitted sets are ignored; instead they
are notionally considered to be all ones (i.e., all capabilities
enabled).

2. We'll dual publish image tags with the pattern `$VERSION-setcap` or similar that have the `m3dbnode` binary with the file capability using `capset cap_sys_resource=+ep /bin/m3dbnode`. This will allow users running as non-root that _do_ give the container `SYS_RESOURCE` to raise the fd limit. We know this approach should work given in 1 release we did add setcap (before we knew the implications) and we can test against that:

```
$ TMP=$(mktemp -d); sudo chown -R 1000:1000 $TMP; docker run --rm -u 1000:1000 --cap-add SYS_RESOURCE -v $TMP:/var/lib/m3db quay.io/m3db/m3dbnode:v0.9.5
2019/06/14 19:55:42 Go Runtime version: go1.10.8
2019/06/14 19:55:42 Build Version:      v0.9.5
2019/06/14 19:55:42 Build Revision:     4d4c1dfc
2019/06/14 19:55:42 Build Branch:       HEAD
2019/06/14 19:55:42 Build Date:         2019-05-28-13:52:53
2019/06/14 19:55:42 Build TimeUnix:     1559051573
{"level":"info","ts":1560542142.140831,"msg":"raised rlimit no file fds limit","required":true,"sysNROpenValue":3000000,"noFileMaxValue":3000000,"noFileCurrValue":3000000}
```

^ non-root user, with `SYS_RESOURCE`, able to run m3db.

3. The operator will add `SYS_RESOURCE` if the user doesn't specify a security context.

4. We'll document this in m3db docs and the operator docs so users are aware of their options and the tradeoffs.

I'm working on PRs for this now.
